File: whati8/api/routers/photo.py
"""Photo recognition API endpoints."""

import logging
from fastapi import APIRouter, Depends, UploadFile, File, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from whati8.api.deps import get_current_user, get_db
from whati8.models.user import User
from whati8.services.photo_recognition import recognize_food_photo

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/photo", tags=["photo"])

MAX_FILE_SIZE = 10 * 1024 * 1024  # 10MB


@router.post("/recognize")
async def recognize_photo(
    file: UploadFile = File(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Upload a food photo for AI recognition."""
    if not file.content_type or not file.content_type.startswith("image/"):
        raise HTTPException(status_code=400, detail="File must be an image")

    image_data = await file.read()
    if len(image_data) > MAX_FILE_SIZE:
        raise HTTPException(status_code=400, detail="Image too large (max 10MB)")

    try:
        result = await recognize_food_photo(
            image_data=image_data,
            media_type=file.content_type,
        )
    except Exception as e:
        logger.error(f"Photo recognition failed: {e}")
        raise HTTPException(status_code=500, detail=f"Recognition failed: {str(e)}")

    return {
        "is_nutrition_label": result.is_nutrition_label,
        "items": [
            {
                "name": item.name,
                "serving_description": item.serving_description,
                "serving_size_g": item.serving_size_g,
                "confidence": item.confidence,
                "nutrients": item.nutrients,
            }
            for item in result.items
        ],
    }

File: frontend/src/lib/api/photo.ts
import { get } from 'svelte/store';
import { authStore } from '../stores/auth';

export interface RecognizedItem {
  name: string;
  serving_description: string;
  serving_size_g: number;
  confidence: string;
  nutrients: Record<string, number>;
}

export interface RecognitionResult {
  is_nutrition_label: boolean;
  items: RecognizedItem[];
}

export async function recognizePhoto(file: File): Promise<RecognitionResult> {
  const { token } = get(authStore);
  const formData = new FormData();
  formData.append('file', file);

  const response = await fetch('/photo/recognize', {
    method: 'POST',
    headers: {
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: formData,
  });

  if (!response.ok) {
    const err = await response.json().catch(() => ({ detail: 'Recognition failed' }));
    throw new Error(err.detail || `HTTP ${response.status}`);
  }

  return response.json();
}
